File: exercises/C/exercise2a/template/mandelbrot.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <error.h>
#include "arralloc.h"
#include "write_ppm.h"
#include "read_options.h"

/* Initialise data for image array.  Data is stored in "scanline
 * order", i.e. x dimension varies fastest.  You get an array with
 * shape image[grid_size_y][grid_size_x] from this function. */
void initialise_image(int ***image, const int grid_size_x, const int grid_size_y)
{
    int i;
    int j;
    *image = (int**)arralloc(sizeof(int), 2, grid_size_y, grid_size_x);

    if ( NULL == *image ) {
        error(1, errno, "Unable to allocate memory for image\n");
    }
    /* initalise results array to black */
    for ( i = 0; i < grid_size_y; i++ ) {
        for ( j = 0; j < grid_size_x; j++ ) {
            (*image)[i][j] = -1;
        }
    }
}

void copy_slice_to_image(int **image_slice, int **image,
                         const int slice, const int nslice,
                         const int grid_size_x, const int grid_size_y)
{
    /*
     * Copy the partial image in IMAGE_SLICE into the global IMAGE.
     *
     * The choice of how the slice number indexes the global image is
     * up to you, but you should be consistent between
     * compute_mandelbrot_slice and copy_slice_to_image.
     */
}

int **compute_mandelbrot_slice(const int slice, const int nslice,
                               const float xmin, const float xmax,
                               const float ymin, const float ymax,
                               const int grid_size_x, const int grid_size_y,
                               const int max_iter)
{
    /*
     * Compute the mandelbrot set in a SLICE of the whole domain.
     * The total number of slices is given by NSLICE.
     *
     * It is your choice whether you slice the domain the x or the y
     * direction, but think about which is going to be more efficient
     * for data access.  Recall the image is stored in scanline order
     * image[grid_size_y][grid_size_x].
     *
     * This function should return a newly initialised image slice.
     */
    return NULL;
}

void compute_mandelbrot_set(int **image,
                            const float xmin,
                            const float xmax,
                            const float ymin,
                            const float ymax,
                            const int grid_size_x,
                            const int grid_size_y,
                            const int max_iter)
{
    int slice;
    int nslice;
    int **image_slice;

    /* Arbitrary number of slices */
    nslice = 3;
    for ( slice = 0; slice < nslice; slice++ ) {
        image_slice = compute_mandelbrot_slice(slice, nslice, xmin, xmax,
                                               ymin, ymax,
                                               grid_size_x, grid_size_y,
                                               max_iter);
        copy_slice_to_image(image_slice, image, slice, nslice,
                            grid_size_x, grid_size_y);
        free(image_slice);
    }
}

int main(int argc, char** argv)
{
    int grid_size_x = GRIDSIZE_X;
    int grid_size_y = GRIDSIZE_Y;
    int max_iter = ITERATIONS;
    float xmin = XMIN;
    float xmax = XMAX;
    float ymin = YMIN;
    float ymax = YMAX;
    int **image;

    read_options(argc, argv, &grid_size_x, &grid_size_y, &max_iter,
                 &xmin, &xmax, &ymin, &ymax);

    initialise_image(&image, grid_size_x, grid_size_y);
    write_ppm("input.ppm", image, grid_size_x, grid_size_y, max_iter);

    compute_mandelbrot_set(image, xmin, xmax, ymin, ymax,
                           grid_size_x, grid_size_y, max_iter);

    /* Sort out colours and write out the output file. */
    write_ppm("output.ppm", image, grid_size_x, grid_size_y, max_iter);

    free(image);
    return 0;
}

File: exercises/F/exercise2a/template/mandelbrot.f90
program mandelbrot

  use write_image
  use read_opt
  implicit none

  integer, allocatable, dimension(:,:) :: image

  integer :: gridsizex, gridsizey
  integer :: iter
  complex :: cmin
  complex :: cmax

  call read_options(gridsizex, gridsizey, iter, cmin, cmax)

  call initialise_image(image, gridsizex, gridsizey)

  call compute_mandelbrot_set(image, cmin, cmax, gridsizex, gridsizey, iter)

  call write_ppm("output.ppm", image, iter)

  deallocate(image)

contains

  ! Initialise the data for the image array.  Data is stored in
  ! "scanline order", i.e. x dimension varies fastest.  You get an
  ! array with shape (grid_size_x, grid_size_y) from this function
  subroutine initialise_image(image, grid_size_x, grid_size_y)
    integer, intent(out), allocatable, dimension(:,:) :: image
    integer, intent(in) :: grid_size_x, grid_size_y

    allocate(image(grid_size_x, grid_size_y))
    image = -1
  end subroutine initialise_image

  subroutine copy_slice_to_image(image_slice, image, &
       slice, nslice)
    integer, dimension(:,:), intent(in) :: image_slice
    integer, dimension(:,:), intent(inout) :: image
    integer, intent(in) :: slice, nslice

    ! Copy the partial imagein IMAGE_SLICE into the global IMAGE.
    !
    ! The choice of how the slice number indexes the global image is
    ! up to you, but you should be consistent between
    ! compute_mandelbrot_slice and copy_slice_to_image.
  end subroutine copy_slice_to_image

  subroutine compute_mandelbrot_slice(image_slice, slice, nslice, &
       min, max, grid_size_x, grid_size_y, max_iter)
    integer, allocatable, dimension(:,:), intent(out) :: image_slice
    integer, intent(in) :: slice, nslice
    complex, intent(in) :: min, max
    integer, intent(in) :: grid_size_x, grid_size_y
    integer, intent(in) :: max_iter

    ! Compute the mandelbrot set in a SLICE of the whole domain.
    ! The total number of slices is given by NSLICE.
    !
    ! It is your choice whether you slice the domain in the x or the
    ! y direction, but think about which is going to be more
    ! efficient for data access.  Recall the image is stored in
    ! scanline order image(grid_size_x, grid_size_y).
    !
    ! This function should allocate the intent(out) variable
    ! image_slice and populate it with the correct values.
  end subroutine compute_mandelbrot_slice

  subroutine compute_mandelbrot_set(image, min, max, grid_size_x, &
       grid_size_y, max_iter)
    integer, dimension(:,:), intent(inout) :: image
    complex, intent(in) :: min, max
    integer, intent(in) :: grid_size_x, grid_size_y
    integer, intent(in) :: max_iter

    integer :: slice
    integer :: nslice
    integer, allocatable, dimension(:,:) :: image_slice

    ! Arbitrary number of slices
    nslice = 3

    do slice = 1, nslice
       call compute_mandelbrot_slice(image_slice, slice, nslice, &
            min, max, grid_size_x, grid_size_y, max_iter)
       call copy_slice_to_image(image_slice, image, slice, nslice)
       if ( allocated(image_slice) ) deallocate(image_slice)
    end do
  end subroutine compute_mandelbrot_set

end program mandelbrot
